File: admin-api/src/main/java/com/tastyhouse/adminapi/review/ReviewService.java
package com.tastyhouse.adminapi.review;

import java.util.List;

import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import com.tastyhouse.core.domain.review.domain.vo.ReviewCommentId;
import com.tastyhouse.core.domain.review.domain.vo.ReviewId;
import com.tastyhouse.core.domain.review.domain.vo.ReviewReplyId;
import com.tastyhouse.core.domain.review.application.ReviewCommandService;
import com.tastyhouse.core.domain.review.application.ReviewQueryService;
import com.tastyhouse.core.domain.review.application.dto.ReviewSearchCondition;
import com.tastyhouse.core.domain.review.application.dto.result.ReviewCommentListItemResult;
import com.tastyhouse.core.domain.review.application.dto.result.ReviewListItemResult;
import com.tastyhouse.core.domain.review.application.dto.result.ReviewManagementDetailResult;
import com.tastyhouse.core.domain.review.application.dto.result.ReviewReplyListItemResult;
import com.tastyhouse.core.exception.EntityNotFoundException;
import com.tastyhouse.core.exception.ErrorCode;
import com.tastyhouse.core.shared.page.PageResult;
import com.tastyhouse.external.file.FileService;
import com.tastyhouse.adminapi.common.PaginationResponse;
import com.tastyhouse.adminapi.review.response.ReviewCommentListItemResponse;
import com.tastyhouse.adminapi.review.response.ReviewListItemResponse;
import com.tastyhouse.adminapi.review.response.ReviewManagementDetailResponse;
import com.tastyhouse.adminapi.review.response.ReviewReplyListItemResponse;

@Service
@RequiredArgsConstructor
public class ReviewService {

    private final ReviewCommandService reviewCommandService;
    private final ReviewQueryService reviewQueryService;
    private final FileService fileService;

    public PaginationResponse<ReviewListItemResponse> getReviews(
        Long shopId,
        Long productId,
        Long memberId,
        Boolean hidden,
        String content,
        Double minRating,
        Double maxRating,
        int page,
        int size
    ) {
        ReviewSearchCondition condition = ReviewSearchCondition.of(shopId, productId, memberId, hidden, content, minRating, maxRating);
        PageResult<ReviewListItemResponse> pageResult = reviewQueryService.findReviews(condition, page, size)
            .map(this::toReviewListItemResponse);
        return PaginationResponse.from(pageResult);
    }

    public ReviewManagementDetailResponse getReview(Long id) {
        ReviewId reviewId = ReviewId.of(id);
        ReviewManagementDetailResult detail = reviewQueryService.findReviewManagementDetail(reviewId)
            .orElseThrow(() -> new EntityNotFoundException(ErrorCode.REVIEW_NOT_FOUND));
        return toReviewManagementDetailResponse(detail);
    }

    public void changeReviewHidden(Long id, boolean hidden) {
        ReviewId reviewId = ReviewId.of(id);
        reviewCommandService.changeReviewHidden(reviewId, hidden);
    }

    public void deleteReview(Long id) {
        ReviewId reviewId = ReviewId.of(id);
        reviewCommandService.deleteReview(reviewId);
    }

    public List<ReviewCommentListItemResponse> getComments(Long id) {
        ReviewId reviewId = ReviewId.of(id);
        List<ReviewCommentListItemResult> comments = reviewQueryService.findCommentsIncludingHidden(reviewId);

        List<ReviewCommentId> commentIds = comments.stream()
            .map(comment -> ReviewCommentId.of(comment.id()))
            .toList();
        List<ReviewReplyListItemResult> replies = reviewQueryService.findRepliesIncludingHidden(commentIds);

        return comments.stream()
            .map(comment -> toReviewCommentListItemResponse(comment, replies))
            .toList();
    }

    public void changeCommentHidden(Long commentId, boolean hidden) {
        ReviewCommentId reviewCommentId = ReviewCommentId.of(commentId);
        reviewCommandService.changeCommentHidden(reviewCommentId, hidden);
    }

    public void deleteComment(Long commentId) {
        ReviewCommentId reviewCommentId = ReviewCommentId.of(commentId);
        reviewCommandService.deleteComment(reviewCommentId);
    }

    public void changeReplyHidden(Long replyId, boolean hidden) {
        ReviewReplyId reviewReplyId = ReviewReplyId.of(replyId);
        reviewCommandService.changeReplyHidden(reviewReplyId, hidden);
    }

    public void deleteReply(Long replyId) {
        ReviewReplyId reviewReplyId = ReviewReplyId.of(replyId);
        reviewCommandService.deleteReply(reviewReplyId);
    }

    private ReviewListItemResponse toReviewListItemResponse(ReviewListItemResult dto) {
        return ReviewListItemResponse.from(
            dto.id(),
            dto.shopId(),
            dto.productId(),
            dto.memberId().value(),
            dto.memberNickname(),
            dto.totalRating(),
            dto.content(),
            dto.hidden(),
            dto.createdAt()
        );
    }

    private ReviewManagementDetailResponse toReviewManagementDetailResponse(ReviewManagementDetailResult dto) {
        return ReviewManagementDetailResponse.from(
            dto.id(),
            dto.shopId(),
            dto.shopName(),
            dto.stationName(),
            dto.content(),
            dto.totalRating(),
            dto.tasteRating(),
            dto.amountRating(),
            dto.priceRating(),
            dto.atmosphereRating(),
            dto.kindnessRating(),
            dto.hygieneRating(),
            dto.willRevisit(),
            dto.hidden(),
            dto.memberId().value(),
            dto.memberNickname(),
            fileService.getUrlByPath(dto.memberProfileImageUrl()),
            dto.createdAt(),
            dto.imageUrls().stream().map(fileService::getUrlByPath).toList(),
            dto.tagNames()
        );
    }

    private ReviewCommentListItemResponse toReviewCommentListItemResponse(ReviewCommentListItemResult comment, List<ReviewReplyListItemResult> replies) {
        List<ReviewReplyListItemResponse> commentReplies = replies.stream()
            .filter(reply -> reply.commentId().equals(comment.id()))
            .map(this::toReviewReplyListItemResponse)
            .toList();

        return ReviewCommentListItemResponse.from(
            comment.id(),
            comment.memberId().value(),
            comment.memberNickname(),
            comment.content(),
            comment.hidden(),
            comment.createdAt(),
            commentReplies
        );
    }

    private ReviewReplyListItemResponse toReviewReplyListItemResponse(ReviewReplyListItemResult dto) {
        return ReviewReplyListItemResponse.from(
            dto.id(),
            dto.memberId().value(),
            dto.memberNickname(),
            dto.replyToMemberId() != null ? dto.replyToMemberId().value() : null,
            dto.replyToMemberNickname(),
            dto.content(),
            dto.hidden(),
            dto.createdAt()
        );
    }
}

File: src/api/review/review.service.ts
import "server-only";

import type { ApiPageRequest, ApiResponse } from "@/api/shared/types";
import type { ReviewComment, ReviewDetail, ReviewListItem } from "@/feature/review/domain";

import type { ReviewListQueryRequest } from "./review.dto";
import { reviewRepository } from "./review.repository";

export const reviewService = {
  // 리뷰 목록 조회 — 도메인 반환
  async getReviews(query: ReviewListQueryRequest, pageRequest: ApiPageRequest): Promise<ApiResponse<ReviewListItem[]>> {
    const res = await reviewRepository.getList(query, pageRequest);
    return {
      ...res,
      data: res.data?.map((item) => ({
        id: item.id,
        shopId: item.shopId,
        productId: item.productId,
        memberId: item.memberId,
        memberNickname: item.memberNickname,
        totalRating: item.totalRating,
        content: item.content,
        hidden: item.hidden,
        createdAt: item.createdAt,
      })),
    };
  },

  // 리뷰 상세 조회 — 도메인 반환
  async getReview(id: number): Promise<ApiResponse<ReviewDetail>> {
    const res = await reviewRepository.getDetail(id);
    if (!res.data) return { ...res, data: undefined };
    return {
      ...res,
      data: {
        id: res.data.id,
        shopId: res.data.shopId,
        shopName: res.data.shopName,
        stationName: res.data.stationName,
        content: res.data.content,
        totalRating: res.data.totalRating,
        tasteRating: res.data.tasteRating,
        amountRating: res.data.amountRating,
        priceRating: res.data.priceRating,
        atmosphereRating: res.data.atmosphereRating,
        kindnessRating: res.data.kindnessRating,
        hygieneRating: res.data.hygieneRating,
        willRevisit: res.data.willRevisit,
        hidden: res.data.hidden,
        memberId: res.data.memberId,
        memberNickname: res.data.memberNickname,
        memberProfileImageUrl: res.data.memberProfileImageUrl,
        createdAt: res.data.createdAt,
        imageUrls: res.data.imageUrls,
        tagNames: res.data.tagNames,
      },
    };
  },

  // 리뷰 댓글/답글 조회 — 도메인 반환
  async getComments(id: number): Promise<ApiResponse<ReviewComment[]>> {
    const res = await reviewRepository.getComments(id);
    return {
      ...res,
      data: res.data?.map((comment) => ({
        id: comment.id,
        memberId: comment.memberId,
        memberNickname: comment.memberNickname,
        content: comment.content,
        hidden: comment.hidden,
        createdAt: comment.createdAt,
        replies: comment.replies.map((reply) => ({
          id: reply.id,
          memberId: reply.memberId,
          memberNickname: reply.memberNickname,
          replyToMemberId: reply.replyToMemberId,
          replyToMemberNickname: reply.replyToMemberNickname,
          content: reply.content,
          hidden: reply.hidden,
          createdAt: reply.createdAt,
        })),
      })),
    };
  },
};
